log4j-api/src/main/java/org/apache/logging/log4j/spi/LoggerRegistry.java

```diff
@@ -87,97 +114,224 @@ public void putIfAbsent(final Map<String, T> innerMap, final String name, final
-        return getOrCreateInnerMap(factoryKey(messageFactory)).containsKey(name);
+    public boolean hasLogger(final String name, @Nullable final MessageFactory messageFactory) {
+        requireNonNull(name, "name");
+        final T logger = getLogger(name, messageFactory);
+        return logger != null;
     }
 
     /**
-     * Detects if a Logger with the specified name and MessageFactory type exists.
-     * @param name The Logger name to search for.
-     * @param messageFactoryClass The message factory class to search for.
-     * @return true if the Logger exists, false otherwise.
+     * Checks if a logger associated with the given name and message factory type exists.
+     *
+     * @param name a logger name
+     * @param messageFactoryClass a message factory class
+     * @return {@code true}, if the logger exists; {@code false} otherwise.
      * @since 2.5
      */
     public boolean hasLogger(final String name, final Class<? extends MessageFactory> messageFactoryClass) {
-        return getOrCreateInnerMap(factoryClassKey(messageFactoryClass)).containsKey(name);
+        requireNonNull(name, "name");
+        requireNonNull(messageFactoryClass, "messageFactoryClass");
+        readLock.lock();
+        try {
+            return loggerByMessageFactoryByName.getOrDefault(name, Collections.emptyMap()).keySet().stream()
+                    .anyMatch(messageFactory -> messageFactoryClass.equals(messageFactory.getClass()));
+        } finally {
+            readLock.unlock();
+        }
+    }
+
+    /**
+     * Registers the provided logger using the given name – <b>message factory parameter is ignored</b> and the one from the logger will be used instead.
+     *
+     * @param name a logger name
+     * @param messageFactory ignored – kept for backward compatibility
+     * @param logger a logger instance
+     * @deprecated As of version {@code 2.24.1}, planned to be removed!
+     * Use {@link #computeIfAbsent(String, MessageFactory, BiFunction)} instead.
+     */
+    @Deprecated
+    public void putIfAbsent(final String name, @Nullable final MessageFactory messageFactory, final T logger) {
+
+        // Check arguments
+        requireNonNull(name, "name");
+        requireNonNull(logger, "logger");
+
+        // Insert the logger
+        writeLock.lock();
+        try {
+            final Map<MessageFactory, T> loggerByMessageFactory =
+                    loggerByMessageFactoryByName.computeIfAbsent(name, this::createLoggerByMessageFactoryMap);
+            final MessageFactory loggerMessageFactory = logger.getMessageFactory();
+            loggerByMessageFactory.putIfAbsent(loggerMessageFactory, logger);
+        } finally {
+            writeLock.unlock();
+        }
+    }
+
+    public T computeIfAbsent(
+            final String name,
+            final MessageFactory messageFactory,
+            final BiFunction<String, MessageFactory, T> loggerSupplier) {
```

Comment:
I get the idea: you don't want multiple threads to create the same logger at the same time and **then** register it with the map.

Since all loggers with the same name and message factory are functionally equivalent, I don't think we really need `computeIfAbsent`, `putIfAbsent` is good enough here.

Reply:
No, that is not the point – even though that is indeed a positive side effect. The reason I introduced `computeIfAbsent()` is:

1. It doesn't allow null `MF` and returns the effective `Logger`. These two cannot be implemented in a backward compatible manner using `putIfAbsent()`.
2. All existing `LC` implementations were doing: `var l = getLogger(); if (l == null) { putIfAbsent(newLogger); } return getLogger();` There were two redundant `getLogger()` calls – since `putIfAbsent()` is not returning the effective logger. `computeIfAbsent()` is not only more performant, but also results in less boilerplate. This is visible in the carried out `LC` changes too.